File: app/src/main/java/io/github/xalrk/nudge/ui/RandomScreen.kt
package io.github.xalrk.nudge.ui

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.setValue
import androidx.compose.foundation.Canvas
import androidx.compose.material3.LocalContentColor
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import kotlin.random.Random
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.MotionDurationScale
import androidx.compose.ui.graphics.graphicsLayer
import kotlinx.coroutines.launch
import kotlin.math.PI
import kotlin.math.sin
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.github.xalrk.nudge.data.FrequencyMode
import io.github.xalrk.nudge.data.Settings

@Composable
fun RandomScreen(vm: NudgeViewModel, onAdd: () -> Unit, onOpen: (Long) -> Unit) {
    val reminders by vm.reminders.collectAsStateWithLifecycle()
    val settings by vm.settings.collectAsStateWithLifecycle()
    val random = reminders.filter { it.isRandom }
    val enabledCount = random.count { it.enabled }

    val overallMean = when (settings.frequencyMode) {
        FrequencyMode.PER_REMINDER -> if (enabledCount > 0) settings.meanIntervalMillis / enabledCount else 0L
        FrequencyMode.WHOLE_POOL -> settings.meanIntervalMillis
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Random reminders") },
                actions = {
                    // Roll: the die spins in place around two axes, flicks through faces, lands on a random one.
                    val spin = remember { Animatable(0f) }
                    var face by remember { mutableIntStateOf(Random.nextInt(1, 7)) }
                    val scope = rememberCoroutineScope()
                    IconButton(onClick = {
                        vm.rerollRandom()
                        scope.launch(object : MotionDurationScale { override val scaleFactor: Float get() = 1f }) {
                            val flicker = List(14) { Random.nextInt(1, 7) }
                            val landing = Random.nextInt(1, 7)
                            spin.snapTo(0f)
                            spin.animateTo(1f, tween(1000, easing = FastOutSlowInEasing)) {
                                face = if (value < 0.8f) flicker[(value / 0.8f * flicker.size).toInt().coerceIn(0, flicker.lastIndex)] else landing
                            }
                        }
                    }) {
                        // spin.value is read inside the layer block, so each frame only updates the
                        // GPU transform of this one layer; nothing recomposes or redraws until the face changes.
                        DieFace(
                            face,
                            Modifier.size(24.dp).graphicsLayer {
                                val t = spin.value
                                rotationY = t * 720f
                                rotationX = t * 360f
                                cameraDistance = 8f * density
                                val s = 1f + 0.15f * sin(t * PI.toFloat())
                                scaleX = s; scaleY = s
                            },
                        )
                    }
                },
            )
        },
        floatingActionButton = { FloatingActionButton(onClick = onAdd, containerColor = MaterialTheme.colorScheme.primary, contentColor = MaterialTheme.colorScheme.onPrimary) { Icon(Icons.Filled.Add, contentDescription = "Add random reminder") } },
    ) { padding ->
        LazyColumn(Modifier.fillMaxSize().padding(padding)) {
            item {
                Column(Modifier.padding(16.dp)) {
                    val hours = "between ${hourLabel(settings.activeStartHour)} and ${hourLabel(settings.activeEndHour)}"
                    val rate = Settings.describeInterval(settings.meanIntervalMillis)
                    val text = when (settings.frequencyMode) {
                        FrequencyMode.PER_REMINDER -> "Each of these fires at an unpredictable moment $hours, $rate on average. Both are configurable in Settings."
                        FrequencyMode.WHOLE_POOL -> "One of these fires at an unpredictable moment $hours, $rate on average. Both are configurable in Settings."
                    }
                    Text(text, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    if (enabledCount > 1 && settings.frequencyMode == FrequencyMode.PER_REMINDER) {
                        Text("With $enabledCount enabled, that adds up to ${Settings.describeInterval(overallMean.coerceAtLeast(Settings.MIN_MEAN_MILLIS))} overall.",
                            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.padding(top = 4.dp))
                    }
                }
            }
            if (random.isEmpty()) item {
                Text("No random reminders yet. Tap + or import a CSV file from Settings.",
                    Modifier.padding(16.dp), color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            items(random, key = { it.id }) { r ->
                val sub = when {
                    !r.enabled -> "Paused"
                    r.nextAt == null -> "Waiting"
                    settings.showNextRandomTime -> "Next " + Fmt.relative(r.nextAt) + " · " + Fmt.instant(r.nextAt).format(Fmt.dayTime)
                    else -> "Sometime soon"
                }
                ReminderRow(r, subtitle = sub, color = vm.colorOf(r), onClick = { onOpen(r.id) }, onToggle = { vm.setEnabled(r.id, it) })
            }
            item { Box(Modifier.size(80.dp)) }
        }
    }
}

/** A die face drawn in the current content color: rounded outline plus 1..6 pips. */
@Composable
private fun DieFace(face: Int, modifier: Modifier = Modifier) {
    val color = LocalContentColor.current
    Canvas(modifier.semantics { contentDescription = "Re-roll times" }) {
        val w = size.width
        val stroke = w * 0.09f
        drawRoundRect(color, topLeft = Offset(stroke / 2, stroke / 2), size = Size(w - stroke, w - stroke),
            cornerRadius = CornerRadius(w * 0.22f), style = Stroke(stroke))
        val c = w / 2
        val o = w * 0.24f
        val pips: List<Offset> = when (face.coerceIn(1, 6)) {
            1 -> listOf(Offset(c, c))
            2 -> listOf(Offset(c - o, c - o), Offset(c + o, c + o))
            3 -> listOf(Offset(c - o, c - o), Offset(c, c), Offset(c + o, c + o))
            4 -> listOf(Offset(c - o, c - o), Offset(c + o, c - o), Offset(c - o, c + o), Offset(c + o, c + o))
            5 -> listOf(Offset(c - o, c - o), Offset(c + o, c - o), Offset(c, c), Offset(c - o, c + o), Offset(c + o, c + o))
            else -> listOf(Offset(c - o, c - o), Offset(c + o, c - o), Offset(c - o, c), Offset(c + o, c), Offset(c - o, c + o), Offset(c + o, c + o))
        }
        for (p in pips) drawCircle(color, radius = w * 0.085f, center = p)
    }
}

fun hourLabel(h: Int): String = when {
    h == 0 || h == 24 -> "midnight"
    h == 12 -> "noon"
    h < 12 -> "${h} am"
    else -> "${h - 12} pm"
}

File: ios/Nudge/Sources/UI/RandomScreen.swift
import SwiftUI
import NudgeCore

struct RandomScreen: View {
    @EnvironmentObject private var engine: ReminderEngine
    @EnvironmentObject private var store: Store
    @EnvironmentObject private var settings: SettingsStore
    @EnvironmentObject private var router: Router
    @Environment(\.theme) private var theme

    // Roll: a real cube tumbles around two axes and eases onto one flat, random face.
    @State private var fromAngles = anglesFor(Int.random(in: 1...6))
    @State private var toAngles = anglesFor(1)
    @State private var turns: (Double, Double) = (0, 0)
    @State private var spin: Double = 0

    var body: some View {
        let s = settings.snapshot
        let random = store.state.reminders.filter { $0.isRandom }
        let enabledCount = random.filter { $0.enabled }.count
        let perReminderMean: Int64 = enabledCount > 0 ? s.meanIntervalMillis / Int64(enabledCount) : 0
        let overallMean: Int64 = s.frequencyMode == .perReminder ? perReminderMean : s.meanIntervalMillis
        Screen {
            ZStack(alignment: .bottomTrailing) {
                ScrollView {
                    LazyVStack(alignment: .leading, spacing: 0) {
                        PauseBanner(settings: s) { engine.setPausedUntil(0) }
                        VStack(alignment: .leading, spacing: 4) {
                            Text(headline(s)).font(.subheadline).foregroundColor(theme.onSurfaceVariant)
                            if enabledCount > 1 && s.frequencyMode == .perReminder {
                                Text(sumLine(enabledCount, overallMean)).font(.footnote).foregroundColor(theme.onSurfaceVariant)
                            }
                        }
                        .padding(16)
                        if random.isEmpty {
                            Text("No random reminders yet. Tap + or import a CSV file from Settings.").foregroundColor(theme.onSurfaceVariant).padding(16)
                        }
                        ForEach(random) { r in
                            ReminderRow(r: r, subtitle: subtitle(r, s), color: engine.colorOf(r), onTap: { router.edit = EditRequest(id: r.id) }, onToggle: { engine.setEnabled(r.id, $0) })
                        }
                        Color.clear.frame(height: 88)
                    }
                }
                Fab(systemImage: "plus", label: "Add random reminder") { router.edit = EditRequest(kind: .random) }
            }
        }
        .navigationTitle("Random Reminders")
        .toolbar {
            ToolbarItem(placement: .navigationBarTrailing) {
                Button(action: roll) {
                    let rx = fromAngles.0 + (toAngles.0 + turns.0 - fromAngles.0) * spin
                    let ry = fromAngles.1 + (toAngles.1 + turns.1 - fromAngles.1) * spin
                    Die3D(rotX: rx, rotY: ry).frame(width: 28, height: 28)
                }
                .accessibilityLabel("Re-roll times")
            }
        }
        .onAppear { toAngles = fromAngles }
    }

    private func headline(_ s: SettingsSnapshot) -> String {
        let hours = "between \(SettingsMath.hourLabel(s.activeStartHour)) and \(SettingsMath.hourLabel(s.activeEndHour))"
        let rate = SettingsMath.describeInterval(s.meanIntervalMillis)
        let who = s.frequencyMode == .perReminder ? "Each" : "One"
        return "\(who) of these fires at an unpredictable moment \(hours), \(rate) on average. Both are configurable in Settings."
    }

    private func sumLine(_ enabledCount: Int, _ overallMean: Int64) -> String {
        "With \(enabledCount) enabled, that adds up to \(SettingsMath.describeInterval(max(overallMean, SettingsMath.minMeanMillis))) overall."
    }

    private func subtitle(_ r: Reminder, _ s: SettingsSnapshot) -> String {
        var rate = ""
        if let m = r.meanOverrideMillis { rate += " · " + SettingsMath.describeInterval(m) }
        if let d = r.randomDaysOrNil() {
            let names = DayOfWeek.allCases.filter { d.contains($0) }.map { String($0.name.prefix(2)).capitalized }
            rate += " · " + names.joined(separator: " ")
        }
        if !r.enabled { return "Paused" }
        guard let n = r.nextAt else { return "Waiting" }
        if s.showNextRandomTime { return "Next " + Fmt.relative(n) + " · " + Fmt.dayTime(n) + rate }
        return "Sometime soon" + rate
    }

    private func roll() {
        engine.rerollRandom()
        let landing = Int.random(in: 1...6)
        // Continue from wherever the die is now so rapid taps never jump.
        let t = spin
        fromAngles = (fromAngles.0 + (toAngles.0 + turns.0 - fromAngles.0) * t, fromAngles.1 + (toAngles.1 + turns.1 - fromAngles.1) * t)
        toAngles = anglesFor(landing)
        turns = (360 * Double(Int.random(in: 1...2)), 360 * Double(Int.random(in: 2...3)))
        spin = 0
        withAnimation(.timingCurve(0.4, 0, 0.2, 1, duration: 1.1)) { spin = 1 }
    }
}
